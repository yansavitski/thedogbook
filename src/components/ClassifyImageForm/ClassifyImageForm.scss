@import 'src/global';

.ClassifyImageForm {
  align-items: center;
  display: flex;
  flex-direction: column;

  &-insertForm {
    margin-top: $baseOffset;
  }

  &-preview {
    margin-top: $baseOffset * 2;
  }
}
