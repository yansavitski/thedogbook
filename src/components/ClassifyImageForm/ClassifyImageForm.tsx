import './ClassifyImageForm.scss'

import classNames from 'classnames'
import { InsertUrlForm } from 'components/InsertUrlForm/InsertUrlForm'
import { Loading } from 'components/Loading/Loading'
import { PreviewImage } from 'components/PreviewImage/PreviewImage'
import { UploadForm } from 'components/UploadForm/UploadForm'
import React, { ChangeEvent, useCallback, useState } from 'react'
import { useDispatch } from 'react-redux'
import {
  setClassifyDogFailed,
  setDogPredictionBreed,
} from 'store/dogClassify/actions'
import { Prediction, useMobilenet } from 'utils/Hooks/useMobilenet'

interface ClassifyImageFormProps {
  className?: string
  disabled?: boolean
}

export const ClassifyImageForm: React.FunctionComponent<ClassifyImageFormProps> = React.memo(
  ({ className }: ClassifyImageFormProps) => {
    const [prediction, setPrediction] = useState<string>()
    const [imageUrl, setImageUrl] = useState<string>()
    const model = useMobilenet()
    const hasModel = Boolean(model)
    const dispatch = useDispatch()
    const handleLoadFile = useCallback(
      (event: ChangeEvent<HTMLInputElement>) => {
        const url = window.URL.createObjectURL(event.target.files?.[0])

        setImageUrl(url)
      },
      [setImageUrl],
    )
    const handleImageClassify = useCallback(
      (event: ChangeEvent<HTMLImageElement>) => {
        if (hasModel) {
          model
            .classify(event.target, 1)
            .then((predictions: Prediction[]) => {
              setPrediction(predictions[0].className)

              dispatch(setDogPredictionBreed(predictions[0].className))
            })
            .catch(() => {
              dispatch(setClassifyDogFailed)
            })
        }
      },
      [model],
    )
    const handleInsertUrl = useCallback(
      (url: string) => {
        setImageUrl(url)
      },
      [setImageUrl],
    )

    return (
      <section className={classNames('ClassifyImageForm', className)}>
        {hasModel ? (
          <>
            <UploadForm
              className="ClassifyImageForm-uploadForm"
              onLoadFile={handleLoadFile}
              disabled={!hasModel}
            />

            <InsertUrlForm
              className="ClassifyImageForm-insertForm"
              onSubmit={handleInsertUrl}
              disabled={!hasModel}
            />

            <PreviewImage
              className="ClassifyImageForm-preview"
              url={imageUrl}
              onLoad={handleImageClassify}
              prediction={prediction}
            />
          </>
        ) : (
          <Loading />
        )}
      </section>
    )
  },
)

ClassifyImageForm.displayName = 'ClassifyImageForm'
