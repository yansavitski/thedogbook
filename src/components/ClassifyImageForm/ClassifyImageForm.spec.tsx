import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { ClassifyImageForm } from './ClassifyImageForm'

describe('<ClassifyImageForm />', () => {
  const middlewares = [thunk]
  const mockStore = configureMockStore(middlewares)

  it('renders correctly', () => {
    expect.assertions(1)
    const tree = shallow(
      <Provider store={mockStore({})}>
        <ClassifyImageForm />
      </Provider>,
    ).dive()

    expect(toJson(tree)).toMatchSnapshot()
  })

  it('renders correctly with optional props', () => {
    expect.assertions(1)
    const tree = shallow(
      <Provider store={mockStore({})}>
        <ClassifyImageForm className="customClassWrapper" disabled />
      </Provider>,
    ).dive()

    expect(toJson(tree)).toMatchSnapshot()
  })
})
