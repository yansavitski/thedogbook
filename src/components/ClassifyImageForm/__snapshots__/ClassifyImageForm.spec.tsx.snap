// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClassifyImageForm /> renders correctly 1`] = `<Memo() />`;

exports[`<ClassifyImageForm /> renders correctly with optional props 1`] = `
<Memo()
  className="customClassWrapper"
  disabled={true}
/>
`;
